README.md
Advance Weather Assistant 
Feb 2022 - Apr 2022

VIDEO LINK https://youtu.be/-DsP7WlibCc
In a collaborative team setting, I spearheaded the development of an Advanced Weather Assistant for a Hackathon event (Dexterix 3.0). This project involved integrating a voice assistant, sensors connected to a Raspberry Pi, and an interactive monitoring website. Coordinating and facilitating the implementation of the project was a key role I played. Leveraging my expertise in Python programming with Flask and frameworks, I optimized the functionality of the system. My proficiency in IoT, Raspberry Pi, and sensor integration was instrumental in achieving this. Effective communication with team members was a priority throughout the project. I streamlined processes, resolved technical challenges, and ensured the project's success. Collaborating with the team, providing mentorship, and executing planned tasks were also part of my responsibilities.


Install Requirement.txt

Run Assistant python Assistant\assistant.py
Run main application python flask\main.py
open in browser 
http://127.0.0.1:5000/

or 

Open START.bat script


flask/news.txt
NoneNoneNoneNoneNoneNoneNoneNoneNoneNoneNoneNoneNoneNoneNoneNone

flask/main.py
from flask import Flask,render_template, url_for
from ssl import ALERT_DESCRIPTION_UNRECOGNIZED_NAME

import time
import datetime
import os
import python_weather
import asyncio
import requests
from bs4 import BeautifulSoup

app=Flask(__name__,template_folder='template')
timeex = time.strftime("%Y-%m-%d  %H:%S")
timee = time.strftime("%H:%S")
datee=time.strftime("%Y-%m-%d")




@app.route("/")
def home():
    
    #y=str(getweather())
    

    with open("C:\\Users\\anura\\Desktop\\putty.log","r") as f:
        
        for line in f:
            pass
            last_line=f.readline()
        first_line = line
    tempp=str(first_line)
    humm=str(last_line)

    


    
    return render_template('weather.html',  timee=timee,datee=datee,temp=tempp,humm=humm)
    
    
    


if __name__ == '__main__':
        
    app.run(debug=True)
    
    

   

flask/intro_ai.py
#import audioop
import pyttsx3
import webbrowser
import datetime
import speech_recognition as sr
import wikipedia
import webbrowser
import time
#import keys
#import convo
import python_weather
import asyncio
import requests
from bs4 import BeautifulSoup
import random
#extra
from selenium import webdriver





#import brian2

engine = pyttsx3.init('sapi5')
voices = engine.getProperty('voices')
#print(voices[1].id)
engine.setProperty('voice', voices[0].id)

# Set Rate
#engine.setProperty('rate', 190)
# Set Volume
#engine.setProperty('volume', 1.0)
# Set Voice (Female)

#brian2.test()

def speak(audio):
    engine.say(audio)
    engine.runAndWait()
def speak2(audio):
    engine.say(audio)
    engine.runAndWait()




def greetme():
   
    hour = int(datetime.datetime.now().hour)
    if hour >= 6 and hour < 12:
        speak("Good Morning Sir")
    elif hour >= 12 and hour < 16:
        speak("Good afternoon Sir")
    elif hour >= 16 and hour < 19:
        speak("Good Evening Sir ")
        
   
        
    #speak("I am Not Jarvis. How may I assist you Sir!!")
    #speak("Sir!!")
def intro():
    speak("I'm your personal weather assistant. Currently showing complete weather information")
    speak("And if you need anything Just ask")

async def getweather():
    # declare the client. format defaults to metric system (celcius, km/h, etc.)
    client = python_weather.Client(format=python_weather.METRIC)

    # fetch a weather forecast from a city
    weather = await client.find("Greater Noida")

    # returns the current day's forecast temperature (int)
    x=print(weather.current.temperature)
    speak("The Weather in Grater Noida is")
    speak(weather.current.temperature)
    speak("degree celcius")
    

    # get the weather forecast for a few days
    

    # close the wrapper once done
    await client.close()




if __name__=="__main__" :
    greetme()
    speak("Hello Sir,   I am your  personal weather assistant. How may i help you.  if you need anything, Just ask")
    

    

flask/static/app.js
let cities = document.querySelectorAll(".loc");
let srch_loc = document.querySelector(".srch_location");
let srch_btn = document.querySelector(".btn_srch");
let currentCity = document.querySelector(".currCity");
let temper = document.querySelector(".temperature");
let cloudy = document.querySelector("#cstatus");
let humidity = document.querySelector("#hstatus");
let wind = document.querySelector("#wstatus");
let condition = document.querySelector(".condition");
let dateput = document.querySelector(".dateput");
let timeput = document.querySelector(".timeput");
let icon = document.querySelector(".icon");

let current_loc = "New Delhi";

cities.forEach((city) => {
  city.addEventListener("click", (e) => {
    current_loc = e.target.innerText;
    fetchWeatherData();
  });
});

srch_btn.addEventListener("click", (e) => {
  if (srch_loc.value.length == 0) {
    alert("Enter any location in input box!");
  } else {
    current_loc = srch_loc.value;
    fetchWeatherData();
    srch_loc.value = "";
  }
  e.preventDefault();
});

function dayOfTheWeek(day, month, year) {
  const weekday = [
    "Sunday",
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday",
  ];
  const dt = new Date(day / month / year).getDay();
  return weekday[dt];
}

function redirect() {
  if(document.getElementById("toggle").checked)
  window.location.href="web2.html";
  else
  window.location.href="app.html";

  document.getElementById("toggle").checked=false;
}

function fetchWeatherData() {
  fetch(
    `http://api.weatherapi.com/v1/current.json?key=14c2e437a3aa4af68f9155229222202&q=${current_loc}&aqi=no`
  )
    .then(function (response) {
      let data = response.json();
      return data;
    })
    .then(function (data) {
      console.log(data);

      currentCity.innerText = data.location.name;
      temper.innerHTML = data.current.temp_c + "&#176" + "C";

      cloudy.innerHTML = data.current.cloud + "%";
      humidity.innerHTML = data.current.humidity + "%";
      wind.innerHTML = data.current.wind_kph + "km/h";
      condition.innerHTML = data.current.condition.text;

      const iconId = data.current.condition.icon.substr(
        "//cdn.weatherapi.com/weather/64x64/".length
      ); // day/143.png
      icon.src = "./icons/" + iconId;

      const date = data.location.localtime;
      const y = parseInt(date.substr(0, 4));
      const m = parseInt(date.substr(5, 2));
      const d = parseInt(date.substr(8, 2));
      const time = date.substr(11);
      dateput.innerHTML = `${dayOfTheWeek(d, m, y)} - ${d}/${m}/${y}`;
      timeput.innerHTML = time;

      let timeOfDay = "day";
      const code = data.current.condition.code;

      if (!data.current.is_day) {
        timeOfDay = "night";
      }

      if (code == 1000) {
        document.body.style.backgroundImage = `url(./images/${timeOfDay}/clear.jpg)`;

        srch_btn.style.background = "#e5ba92";
        if (timeOfDay == "night") {
          srch_btn.style.background = "#181e27";
        }
      } else if (
        code == 1003 ||
        code == 1006 ||
        code == 1009 ||
        code == 1030 ||
        code == 1069 ||
        code == 1087 ||
        code == 1135 ||
        code == 1273 ||
        code == 1276 ||
        code == 1279 ||
        code == 1282
      ) {
        document.body.style.backgroundImage = `url(./images/${timeOfDay}/cloudy.jpg)`;
        srch_btn.style.background = "#fa6d1b";
        if (timeOfDay == "night") {
          srch_btn.style.background = "#181e27";
        }
      } else if (
        code == 1063 ||
        code == 1069 ||
        code == 1072 ||
        code == 1150 ||
        code == 1153 ||
        code == 1180 ||
        code == 1183 ||
        code == 1189 ||
        code == 1192 ||
        code == 1195 ||
        code == 1204 ||
        code == 1207 ||
        code == 1240 ||
        code == 1243 ||
        code == 1246 ||
        code == 1249 ||
        code == 1252
      ) {
        document.body.style.backgroundImage = `url(./images/${timeOfDay}/rainy.jpg)`;
        srch_btn.style.background = "#647d75";
        if (timeOfDay == "night") {
          srch_btn.style.background = "#325c80";
        }
      } else {
        document.body.style.backgroundImage = `url(./images/${timeOfDay}/snowy.jpg)`;
        srch_btn.style.background = "#4d72aa";
        if (timeOfDay == "night") {
          srch_btn.style.background = "#1b1b1b";
        }
      }
      document.body.style.opacity = "1";
    });
  /*
    .catch(()=>{
        alert('City not found, please try again!');
        document.body.style.opacity="1";
    });
    */
}

fetchWeatherData();
document.body.style.opacity = "1";

flask/assistant.py
#import audioop
import pyttsx3
import webbrowser
import datetime
import speech_recognition as sr
import wikipedia
import webbrowser
import time
#import keys
#import convo
import python_weather
import asyncio
import requests
from bs4 import BeautifulSoup
import random
#extra
from selenium import webdriver





#import brian2

engine = pyttsx3.init('sapi5')
voices = engine.getProperty('voices')
#print(voices[1].id)
engine.setProperty('voice', voices[0].id)

# Set Rate
#engine.setProperty('rate', 190)
# Set Volume
#engine.setProperty('volume', 1.0)
# Set Voice (Female)

#brian2.test()

def speak(audio):
    engine.say(audio)
    engine.runAndWait()




def greetme():

    hour = int(datetime.datetime.now().hour)
    if hour >= 6 and hour < 12:
        speak("Good Morning Sir")
    elif hour >= 12 and hour < 16:
        speak("Good afternoon Sir")
    elif hour >= 16 and hour < 19:
        speak("Good Evening Sir ")
        
        
    #speak("I am Not Jarvis. How may I assist you Sir!!")
    #speak("Sir!!")
def intro():
    speak("I'm your personal weather assistant. Currently showing   complete weather information")
    speak("  And  if you need anything   Just ask")

async def getweather():
    # declare the client. format defaults to metric system (celcius, km/h, etc.)
    client = python_weather.Client(format=python_weather.METRIC)

    # fetch a weather forecast from a city
    weather = await client.find("Greater Noida")

    # returns the current day's forecast temperature (int)
    x=print(weather.current.temperature)
    speak("The Weather in Grater Noida is")
    speak(weather.current.temperature)
    speak("degree celcius")
    

    # get the weather forecast for a few days
    

    # close the wrapper once done
    await client.close()


async def getweatherfull():
    # declare the client. format defaults to metric system (celcius, km/h, etc.)
    client = python_weather.Client(format=python_weather.METRIC)

    # fetch a weather forecast from a city
    weather = await client.find("Greater Noida")

    # returns the current day's forecast temperature (int)
    print(weather.current.temperature)

    # get the weather forecast for a few days
    for forecast in weather.forecasts:
        print(str(forecast.date), forecast.sky_text, forecast.temperature)
        #speak(str(forecast.date), forecast.sky_text, forecast.temperature)

    # close the wrapper once done
    await client.close()



def tellDay():
      
    
    day = datetime.datetime.today().weekday() + 1
      
    #this line tells us about the number 
    # that will help us in telling the day
    Day_dict = {1: 'Monday', 2: 'Tuesday', 
                3: 'Wednesday', 4: 'Thursday', 
                5: 'Friday', 6: 'Saturday',
                7: 'Sunday'}
      
    if day in Day_dict.keys():
        day_of_the_week = Day_dict[day]
        print(day_of_the_week)
        speak("The day is " + day_of_the_week)
  
  
def tellTime():
      
    # This method will give the time
    time = str(datetime.datetime.now())
      
    # the time will be displayed like 
    # this "2020-06-05 17:50:14.582630"
    #nd then after slicing we can get time
    print(time)
    hour = time[11:13]
    min = time[14:16]
    speak("Currently " + hour + "Hours and" + min + "Minutes") 

def tellTime2():
    
    from datetime import datetime
    print(datetime.today().strftime("%I:%M %p"))
    speak(datetime.today().strftime("It's %I:%M %p"))




def technews(): 
    #url='https://www.bbc.com/news'
    url='https://news.google.com/topics/CAAqJggKIiBDQkFTRWdvSUwyMHZNRGRqTVhZU0FtVnVHZ0pWVXlnQVAB?hl=en-IN&gl=IN&ceid=IN:en'
    #topnews=url='https://news.google.com/topstories?hl=en-IN&gl=IN&ceid=IN%3Aen'    
    response = requests.get(url)
    soup = BeautifulSoup(response.text, 'html.parser')
    headlines = soup.find('body').find_all('h3')

    timeout = time.time() + 60 
    for x in headlines:
        if time.time() > timeout:
            break
        print('=>')
        print(x.text.strip())
        speak(x.text.strip())
        """""
        commandnow()
        if 'stop' in query:
            speak('ok')
            break
        else:
            continue
    
        """
        #query=takecommand().lower()
        #if 'stop' in query:
        #    break

def topnews(): 
    #url='https://www.bbc.com/news'
    #tech=url='https://news.google.com/topics/CAAqJggKIiBDQkFTRWdvSUwyMHZNRGRqTVhZU0FtVnVHZ0pWVXlnQVAB?hl=en-IN&gl=IN&ceid=IN:en'
    url='https://news.google.com/topstories?hl=en-IN&gl=IN&ceid=IN%3Aen'    
    response = requests.get(url)
    soup = BeautifulSoup(response.text, 'html.parser')
    headlines = soup.find('body').find_all('h3')

    timeout = time.time() + 20 
    for x in headlines:
        if time.time() > timeout:
            break
        print('=>')
        print(x.text.strip())
        speak(x.text.strip())
    
def worldnews(): 
    #url='https://www.bbc.com/news'
    #tech=url='https://news.google.com/topics/CAAqJggKIiBDQkFTRWdvSUwyMHZNRGRqTVhZU0FtVnVHZ0pWVXlnQVAB?hl=en-IN&gl=IN&ceid=IN:en'
    url='https://news.google.com/topics/CAAqKggKIiRDQkFTRlFvSUwyMHZNRGx1YlY4U0JXVnVMVWRDR2dKSlRpZ0FQAQ?hl=en-IN&gl=IN&ceid=IN%3Aen'    
    response = requests.get(url)
    soup = BeautifulSoup(response.text, 'html.parser')
    headlines = soup.find('body').find_all('h3')

    timeout = time.time() + 20 
    for x in headlines:
        if time.time() > timeout:
            break
        print('=>')
        print(x.text.strip())
        speak(x.text.strip())
def indianews(): 
    #url='https://www.bbc.com/news'
    #tech=url='https://news.google.com/topics/CAAqJggKIiBDQkFTRWdvSUwyMHZNRGRqTVhZU0FtVnVHZ0pWVXlnQVAB?hl=en-IN&gl=IN&ceid=IN:en'
    url='https://news.google.com/topics/CAAqJQgKIh9DQkFTRVFvSUwyMHZNRE55YXpBU0JXVnVMVWRDS0FBUAE?hl=en-IN&gl=IN&ceid=IN%3Aen'    
    response = requests.get(url)
    soup = BeautifulSoup(response.text, 'html.parser')
    headlines = soup.find('body').find_all('h3')

    timeout = time.time() + 20 
    for x in headlines:
        if time.time() > timeout:
            break
        print('=>')
        print(x.text.strip())
        speak(x.text.strip())
    
    
def dailynews():
    speak("Sir Do you want the Daily News ")
    query=takecommand().lower()
    if 'yes' in query:
        technews()
    elif 'no' in query:
        speak('ok')
def showtask():
    speak("No task assigned for today")

def todaytask():

    lst=["Do you want to checkout today's task sir?","let me remind you sir we have some tasks for today do you want to check them now or later"]
    random.shuffle(lst)
    for tasks in lst:
        #print tasks
        speak(tasks)
    #speak("Do you want to checkout today's task sir?")
    query=takecommand().lower()
    if 'yes' in query:
        showtask()
    elif 'no' in query:
        speak('ok')
    


            
  


def takecommand():

    r = sr.Recognizer()
    with sr.Microphone() as source:
        print("Listening...")
        r.pause_threshold = 4
        audio = r.listen(source)
    try:
        print("Recognizing...")    
        query = r.recognize_google(audio, language='en-in') #Using google for voice recognition.h
        print(f"User said: {query}\n")  #User query will be printed.

    except Exception as e:
        # print(e)    
        print("Say that again please...")   #Say that again will be printed in case of improper voice 
        return "None" #None string will be returned
    return query    

#extra
def search_web():
  
    driver = webdriver.Firefox()
    driver.implicitly_wait(1)
    driver.maximize_window()
  
    if 'youtube' in query:
  
        speak("Opening in youtube")
        indx = takecommand.lower().split().index('youtube')
        query=takecommand.split()[indx + 1:]
        #query = input.split()[indx + 1:]
        driver.get("http://www.youtube.com/results?search_query =" + '+'.join(query))
        return
    else:
        speak("unable to find")





if __name__=="__main__" :
        loop = asyncio.get_event_loop()
        loop.run_until_complete(getweather())
        greetme()
        while True:
            query=takecommand().lower()    
            if 'wikipedia' in query:
                speak('Results from Wikipedia Sir!')
                query = query.replace("wikipedia", "")
                results = wikipedia.summary(query, sentences=2) 
                speak("According to Wikipedia")
                print(results)
                speak(results)

            elif 'open youtube' in query:
                webbrowser.open("youtube.com")
            
            elif 'temperature' in query:
                with open("C:\\Users\\anura\\Desktop\\putty.log","r") as f:
                    first_line=f.readline()
                    for line in f:
                        pass
                    last_line = line

                tempp=str(first_line)
                humm=str(last_line)
                speak(tempp)
            elif 'humidity' in query:
                with open("C:\\Users\\anura\\Desktop\\putty.log","r") as f:
                    first_line=f.readline()
                    for line in f:
                        pass
                    last_line = line

                tempp=str(first_line)
                humm=str(last_line)
                speak(humm)

            elif "which day it is" in query:
                tellDay()
            elif "today's day" in query:
                tellDay()            
            elif 'tell me the time' in query:
                tellTime()
            elif 'time right now' in query:
                tellTime()
            elif 'time' in query:
                tellTime()
            elif 'complete weather' in query:
                getweatherfull()
            elif 'yes' in query:
                speak("can you repeat? sir,")
            elif 'date' in query:
                x = datetime.datetime.now()
                print(x.date)
                speak(x.date)
            elif 'weather' in query:
                loop = asyncio.get_event_loop()
                loop.run_until_complete(getweather())
            elif 'who are you' in query:
                speak("Hello Sir, I'm personal weather assistant i displays and tell complete weather information I am your morning star if you need anything Just ask")
            elif 'Introduce yourself'in query:
                speak("Hello Sir, I'm personal weather assistant i displays and tell complete weather information I am your morning star if you need anything Just ask")
            
            elif 'hello' in query:
                greetme()
                tellDay()
                
                
                dailynews()
            elif 'tech news' in query:
                technews()
            elif 'global news' in query:
                worldnews()
            elif 'top news' in query:
                topnews()
            elif 'india news' in query:
                indianews()
                

            elif 'stop' in query:
                exit()

            elif 'jarvis logout' in query:
                speak('sir ? sir? logging off sir')
                exit()

            elif 'hey jarvis' in query:
                speak("Yes Sir!! i'm here waht do you need")
            elif 'hey ladies' in query:
                speak('sir ? Are you kidding, i, am, not, your, ladies, sir!  give me some, respect! please?')